src/psi/j/launchers/script_based_launcher.py

```diff
@@ -0,0 +1,83 @@
+from pathlib import Path
+from typing import Optional, List
+
+from psi.j.launchers.launcher import Launcher
+from psi.j.job_executor_config import JobExecutorConfig
+from psi.j.job import Job
+
+
+def _str(obj: Optional[object]) -> str:
+    if obj is not None:
+        return str(obj)
+    else:
+        return ''
+
+
+class ScriptBasedLauncher(Launcher):
+    """
+    A launcher that uses a script to start the job, possibly by wrapping it in other tools.
+
+    This launcher is an abstract base class for launchers that wrap the job in a script. The script
+    must be a bash script and is invoked with the first four parameters as:
+        - the job ID
+        - a launcher log file, which is taken from the launcher_log_file configuration setting and
+        defaults to `/dev/null`
+        - the pre- and post- launcher scripts, or empty strings if they are not specified
+
+    Additional positional arguments to the script can be specified by subclasses by overriding
+    the :func:`~psi.j.launchers.script_based_launcher.ScriptBasedLauncher._get_additional_args`
+    function.
+
+    The remaining arguments to the script are the job executable and arguments.
+
+    A simple script library is provided in scripts/lib.sh. Its use is optional and it is intended
+    to be included in a main launcher script using a standard C preprocessor. It does the following:
```

Comment:
I am a bit confused why the C preprocessor is needed - couldn't the same be achieved by this just being a shell script?  But I am probably missing something again...  You seem to invoke the preprocessor to build the actual scripts  during install - but you could also just install `lib.sh` and source it at runtime?  Or are you trying to optimize the sourcing away?

Reply:
I'm trying to optimize the sourcing away. This is probably premature optimization.

I am also trying to avoid figuring out where lib.sh is when the parent script is invoked.

There are probably alternative ways to do things here which may be better.

Reply:
> I'm trying to optimize the sourcing away. This is probably premature optimization.

Indeed, I would say it is... :-)

> I am also trying to avoid figuring out where lib.sh is when the parent script is invoked. 
> There are probably alternative ways to do things here which may be better.

Fair point.  Just adding this as input:

we install shell scripts along with the python sources, so the python module always finds the script with 
```py
script = '%s/script.sh` % os.path.dirname(__file__)
```

Probably not foolproof, but it served us well so far.  Another option is to install into `bin`: `.` and `source` [actually search the path](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_18), so it will be found via `$PATH`.

Reply:
> ```python
> script = '%s/script.sh` % os.path.dirname(__file__)
> ```

That's what's happening. But there's also the step of sourcing lib.sh from another .sh, and there is no __file__ at that stage. It's possible to figure it out using $0, but I opted against that for now. 

There's also the soon-to-be-an-issue of submit scripts (e.g., for PBS) and whether launcher scripts would be called from them or inlined into them.